File: Arsemi-Core/IPC/MessageParsing.cs
using System.Threading.Tasks;

namespace Arsemi {

    namespace IPC {

        public class MessageParsing {
            private ArsemiCore _arsemiCore;

            private SerialPackage _queuedPackage = new();

            private Mutex _mutex = new(false);
            private SemaphoreSlim _semaphore = new(0, 1);
            private const int ParsingTimeoutMs = 1000;

            private DateTime _handshakeTimeout = DateTime.MinValue;
            public enum ConnectionResult {
                NONE,
                WAITING,
                SUCCESS,
                PORT_ERROR,
                TIMEOUT,
            }
            private ConnectionResult _handshakeResult = ConnectionResult.NONE;



            public MessageParsing(ArsemiCore arsemiCore) {
                _arsemiCore = arsemiCore;
            }


            /// <summary>
            /// Connects the microcontroller at the specified serial port and waits for 2 seconds to give the microcontroller time to reset.
            /// </summary>
            /// <param name="serialPortInfo"></param>
            /// <param name="drtResetWaitMs"></param>
            /// <param name="timeoutMs"></param>
            /// <returns>SUCCESS if the microcontroller is connected and the port is available, otherwise false</returns>
            public async Task<ConnectionResult> ConnectMicrocontrollerAsync(SerialPortInfo? serialPortInfo = null, int drtResetWaitMs = 2000, int timeoutMs = 5000) {
                serialPortInfo ??= new();
                Console.WriteLine("Try connecting to microcontroller on " + serialPortInfo.PortName + "...");
                SerialMessaging.Begin(serialPortInfo);
                await Task.Delay(drtResetWaitMs);
                SerialMessaging.DataReceivedAction += () => ParseMessage(); // DEBUG -> later move to start loop

                bool connected = SerialMessaging.PortAvailable();
                if(connected) { Console.WriteLine("Successfully connected to microcontroller on " + serialPortInfo.PortName); }
                else {
                    Console.WriteLine("Error connecting to microcontroller on " + serialPortInfo.PortName + "!");
                    return ConnectionResult.PORT_ERROR;
                }

                return await RequestHandshakeAsync(timeoutMs);
            }


            /// <summary>
            /// Send handshake package over serial
            /// PC sends message over port and checks if arduino replies, if yes this is the microcontroller with the Arsemi-Arduino script.
            /// </summary>
            /// <param name="waitTimeMilliseconds">For how long the method will wait for the handshake</param>
            /// <returns>SUCCESS when the connected microcontroller responded in time, WAITING if there was a handshake request already, otherwise TIMEOUT</returns>
            public async Task<ConnectionResult> RequestHandshakeAsync(int waitTimeMilliseconds) {
                if(_handshakeResult == ConnectionResult.WAITING) return ConnectionResult.WAITING;

                SerialPackage requestHandshakePackage = new SerialPackage(SerialProtocol.Action.System.RequestHandshake);
                SerialMessaging.Write(requestHandshakePackage);

                _handshakeResult = ConnectionResult.WAITING;
                _handshakeTimeout = DateTime.Now.AddMilliseconds(waitTimeMilliseconds);
                Console.WriteLine("Waiting for handshake reply...");
                return await WaitForHandshakeReplyAsync();
            }


            /// <summary>
            /// Parse new messages every 50 milliseconds until a new message is received, return if the new message is the handshake
            /// </summary>
            /// <returns>SUCCESS when _handshakeResult is set to true in the ParseMessage() due to the handshake reply, otherwise TIMEOUT</returns>
            private async Task<ConnectionResult> WaitForHandshakeReplyAsync() {
                while(_handshakeTimeout.CompareTo(DateTime.Now) > 0) {
                    if(_handshakeResult == ConnectionResult.SUCCESS) {
                        Console.WriteLine("Received handshake reply!");
                        return ConnectionResult.SUCCESS;
                    }
                }

                Console.WriteLine("Handshake timeout! Try connecting again on a different port.");
                return ConnectionResult.TIMEOUT;
            }


            /// <summary>
            /// Converts a new message from string to package and then matches the action code to the required actions
            /// </summary>
            public async Task ParseMessage() {
                if(await _semaphore.WaitAsync(ParsingTimeoutMs)) {
                    return; // failed to acquire, other thread will handle the message parsing
                }

                try {
                    while(SerialMessaging.AvailableBytes()) {

                        // Parse action code
                        if(_queuedPackage.Empty) {
                            int queuedActionCode = ParseNextActionCode();
                            if(queuedActionCode == -1)
                                break;
                            _queuedPackage.ActionCode = (byte)queuedActionCode;
                        }

                        var k = SerialMessaging.Buffer; // DEBUG
                        if(!_queuedPackage.Done)
                            await ParseNextParametersAsync();

                        bool done = false;

                        switch(_queuedPackage.ActionCode) {
                        /// Codes meant for sending to the microcontroller -> no need to implement
                        // case SerialProtocol.SystemCodes.HibernateMicrocontroller:
                        //   throw new NotImplementedException();
                        // case SerialProtocol.SystemCodes.WakeMicrocontroller:
                        //   throw new NotImplementedException();
                        // case SerialProtocol.SetupCodes.ClearConfiguration:
                        //   throw new NotImplementedException();
                        // case SerialProtocol.SetupCodes.AddSensor:
                        //   throw new NotImplementedException();
                        // case SerialProtocol.SystemCodes.RequestHandshake:
                        //   throw new NotImplementedException();

                        /// Codes meant for receiving from the microcontroller
                        case SerialProtocol.Action.System.Error:
                            done = ParseSystemError();
                            break;

                        case SerialProtocol.Action.System.ReplyHandshake:
                            if(CheckCRC8Checksum(_queuedPackage))
                                _handshakeResult = ConnectionResult.SUCCESS;
                            break;

                        case SerialProtocol.Action.Sensor.NewSample:
                            //Console.WriteLine("New sample");
                            done = ParseNewSample();
                            break;

                        case SerialProtocol.Action.Setup.SuccessfullyAddedSensor:
                            if(_queuedPackage.Parameters.Count != 1)
                                throw new Exception("Error parsing Message...");
                            int param = _queuedPackage.Parameters[0];
                            if(CheckCRC8Checksum(_queuedPackage))
                                Console.WriteLine("Successfully added a sensor on the microcontroller. -> " + param);
                            break;

                        case SerialProtocol.Action.Setup.SuccessfullyClearedConfiguration:
                            if(CheckCRC8Checksum(_queuedPackage))
                                Console.WriteLine("Successfully cleared the configuration on the microcontroller.");
                            break;

                        case SerialProtocol.Action.System.Debug:
                            if(_queuedPackage.Parameters.Count != 1)
                                break;
                            byte debugParam = _queuedPackage.Parameters[0];
                            if(CheckCRC8Checksum(_queuedPackage)) {
                                Console.Write("Debug reached! ");
                                Console.WriteLine(debugParam);
                            }
                            break;

                        case SerialProtocol.Action.System.Heartbeat:
                            if(CheckCRC8Checksum(_queuedPackage))
                                Console.WriteLine("*");
                            break;

                        default:
                            Console.WriteLine("The action code " + _queuedPackage.ActionCode + " in the message can't be associated with a command in " + nameof(SerialProtocol));
                            //throw new NotImplementedException("The action code " + _queuedActionCode + " in the message can't be associated with a command in " + nameof(SerialProtocol));
                            break;
                        }

                        if(_queuedPackage.Done)
                            _queuedPackage.Reset();
                    }
                }
                finally {
                    _queuedPackage.Reset();
                    _semaphore.Release();
                }
            }


            /// <summary>
            /// 
            /// </summary>
            /// <returns></returns>
            private async Task ParseNextParametersAsync() {
                for(int i = 0; ; i++) {
                    int nextByte = SerialMessaging.PeekByte();
                    if(nextByte == SerialProtocol.PackageStartByte) {
                        _queuedPackage.Crc8 = _queuedPackage.Parameters.Last();
                        _queuedPackage.Parameters.RemoveAt(_queuedPackage.Parameters.Count - 1);
                        _queuedPackage.Done = true;
                        SerialMessaging.DequeueByte(); // discard end of package
                        Console.Write("Received: [ 0 | " + _queuedPackage.ActionCode + " | ");
                        for(int j = 0; j < _queuedPackage.Parameters.Count; j++) {
                            Console.Write(_queuedPackage.Parameters[j] + " | ");
                        }
                        Console.WriteLine(_queuedPackage.Crc8 + " | 0]");
                        return;
                    }
                    else if(nextByte != -1) {
                        _queuedPackage.AppendParameter((byte)SerialMessaging.DequeueByte());
                    }
                    await Task.Delay(5);
                }
            }


            /// <summary>
            /// Parses sensorId, value and checksum from the next 3 bytes in the serial stream.
            /// </summary>
            /// <returns>false if the package doesn't have enough bytes, otherwise true (also in case of message corruption to discard the message)</returns>
            private bool ParseNewSample() {
                if(_queuedPackage.Parameters.Count != 2) {
                    throw new Exception("Error parsing package, wrong count of parameters...");
                }

                CheckCRC8Checksum(_queuedPackage);

                byte sensorId = _queuedPackage.Parameters[0];
                byte value = _queuedPackage.Parameters[1];

                Console.Write("Received sensor data from sensorId: " + sensorId + " with a value of: " + value);
                Console.WriteLine(" | Sensorname: " + _arsemiCore.Sensors[sensorId].Data.Name);
                Console.WriteLine("---");

                _arsemiCore.Sensors[sensorId].Data.Value = value;
                _arsemiCore.NewDataReceived?.Invoke(sensorId, value);
                _arsemiCore.Sensors[sensorId].CheckEventsConditions();
                return true;
            }


            /// <summary>
            /// Matches the different sensor errors to their error messages.
            /// </summary>
            /// <return></return>
            private bool ParseSystemError() {
                if(_queuedPackage.Parameters.Count < 1) {
                    throw new Exception("Error parsing package, not enough parameters...");
                }

                byte errorCode = _queuedPackage.Parameters[0];

                switch(errorCode) {
                case SerialProtocol.Error.Package.InvalidActionCode:
                    if(_queuedPackage.Parameters.Count != 3) {
                        return true; // discard package
                    }
                    CheckCRC8Checksum(_queuedPackage);
                    Console.WriteLine("Received error: " + errorCode + " = " + SerialProtocol.TryGetErrorName(errorCode) +
                                        " -> " + _queuedPackage.Parameters[1]);
                    break;
                case SerialProtocol.Error.Package.InvalidChecksum:
                    if(_queuedPackage.Parameters.Count != 3) {
                        return true; // discard package
                    }
                    CheckCRC8Checksum(_queuedPackage);
                    Console.WriteLine("Received error: " + errorCode + " = " + SerialProtocol.TryGetErrorName(errorCode)
                                        + " -> " + _queuedPackage.Parameters[1] + " != " + _queuedPackage.Parameters[2]);
                    break;
                case SerialProtocol.Error.Package.InvalidSensorParameters:
                    Console.WriteLine("Invalid sensor parameter count...");
                    break;
                default:
                    break;
                }

                return true;
            }


            /// <summary>
            /// Discards all bytes from the Serial stream until StartByte is reached.
            /// </summary>
            /// <returns>-1 when the start byte couldn't be parsed because the queue is too short</returns>
            public static int ParseNextActionCode() {
                while(SerialMessaging.AvailableBytes(2)) {
                    if(SerialMessaging.PeekByte() == SerialProtocol.PackageStartByte) {
                        _ = SerialMessaging.DequeueByte(); // discard StartByte
                        int actionCode = SerialMessaging.DequeueByte();
                        return actionCode;
                    }
                    _ = SerialMessaging.DequeueByte(); // discard
                }
                return -1;
            }

            /// <summary>
            /// Check if the next checksum byte in the serial buffer is equal to the computed checksum of the data
            /// </summary>
            /// <param name="data">data of the package, without StartByte and checksum</param>
            /// <returns></returns>
            /// <exception cref="Exception"></exception>
            private static bool CheckCRC8Checksum(SerialPackage package) {
                byte[] serializedPackage = package.Serialize();
                byte computedCrc8Checksum = SerialMessaging.CRC8(serializedPackage);

                if(package.Crc8 != computedCrc8Checksum) {
                    //throw new Exception("HEY! Loss of packages... :c");
                    Console.WriteLine("Loss of packages in action " + SerialProtocol.TryGetActionName(package.ActionCode) + ", checksum is not the same! -> " + package.Crc8 + "(package) != " + computedCrc8Checksum + "(computed)");
                    return false;
                }
                else return true;
            }
        }
    }
}

File: Arsemi-Arduino/src/IPC/MessageParsing.cpp
#include "MessageParsing.h"

MessageParsing::MessageParsing(ArsemiArduinoCore &newArsemiArduinoCore)
    : arsemiArduinoCore(newArsemiArduinoCore) {}

/// @brief Parse a new serial package with its action code to the
/// according actions and invoke associated functions, with all parameters
/// of the package.
void MessageParsing::parseMessage() {
  if (!SerialMessaging::isPackageAvailable())
    return;

  if (_queuedActionCode == -1 || _queuedActionCode == 0) {
    _queuedActionCode = parseNextActionCode();
    // uint8_t pack[2] = {SerialProtocol::Action::System::Debug,
    //                    _queuedActionCode};
    // SerialMessaging::write(pack, 2);
    if (_queuedActionCode == -1 || _queuedActionCode == 0) {
      return;
    }
  }

  if (!queuedPackage.Done)
    parseParameters();

  bool done = false;

  switch (_queuedActionCode) {
  case SerialProtocol::Action::System::RequestHandshake:
    if (checkCrc8Checksum(queuedPackage)) {
      SerialMessaging::write(SerialProtocol::Action::System::ReplyHandshake);
    }
    done = true;
    break;

  case SerialProtocol::Action::System::HibernateMicrocontroller:
    if (checkCrc8Checksum(queuedPackage)) {
      arsemiArduinoCore.execution = false;
    }
    done = true;
    break;

  case SerialProtocol::Action::System::WakeMicrocontroller:
    if (checkCrc8Checksum(queuedPackage)) {
      arsemiArduinoCore.execution = true;
    }
    done = true;
    break;

  case SerialProtocol::Action::Setup::ClearConfiguration:
    if (checkCrc8Checksum(queuedPackage)) {
      SerialMessaging::write(
          SerialProtocol::Action::Setup::SuccessfullyClearedConfiguration);
      arsemiArduinoCore.destroyAllSensors();
    }
    done = true;
    break;

  case SerialProtocol::Action::Setup::AddSensor:
    done = parseAddSensorAction();
    break;

  default:
    uint8_t data[3] = {SerialProtocol::Action::System::Error,
                       SerialProtocol::Error::Package::InvalidActionCode,
                       (uint8_t)_queuedActionCode};
    SerialMessaging::write(data, 3);
    done = true;
  }

  if (done)
    _queuedActionCode = -1;
}

/// @brief Peeks for PackageDelimiter and discards everything until the action
/// code. Packages are only done if they contain [PackageDelimiter + ActionCode
/// + CRC8], thus this method waits for 3 bytes in the stream. Further checks
/// and processing must be done by another method.
/// @returns Action code which follows after the next start byte, otherwise -1
int MessageParsing::parseNextActionCode() {
  while (SerialMessaging::isPackageAvailable()) {
    if (Serial.peek() == SerialProtocol::PackageDelimiter) {
      Serial.read(); // discard PackageDelimiter
      return Serial.read();
    }

    Serial.read(); // discard
  }
  return -1;
}

/// @brief Parse the package parameters from the message into queuedPackage
void MessageParsing::parseParameters() {
  for (int i = queuedPackage.getParameterCount();; i++) {
    int nextByte = Serial.peek();
    if (nextByte == SerialProtocol::PackageDelimiter) {
      queuedPackage.Crc8 = queuedPackage.getLastParameter();
      queuedPackage.removeLastParameters();
      queuedPackage.Done = true;
      Serial.read(); // discard PackageDelimiter at end of package
      return;
    } else if (nextByte != -1) {
      queuedPackage.appendParameters(Serial.read(), 1);
    }
  }
}

/// @brief Parse action "Add Sensor".
/// Package parameters: [ sensorType | intervalMs |
/// constructorParameters[] (optional, different for each sensor type) ]
/// @return false when still waiting for parameters or the parameters are not
/// enough, otherwise true (it will also return true to discard invalid
/// packages)
bool MessageParsing::parseAddSensorAction() {
  if (queuedPackage.getParameterCount() < 2)
    return true; // TODO: Error message (currently only discards the corrupt
                 // package)

  // SerialMessaging::write(SerialProtocol::Action::System::Debug, 69);

  if (queuedSensor == nullptr) {
    queuedSensor = SensorFactory::createNewSensor(
        AbstractSensor::SensorTypes(queuedPackage.getParameter(0)));
    SerialMessaging::write(SerialProtocol::Action::System::Debug,
                           queuedPackage.getParameter(0));

    if (queuedSensor == nullptr)
      SerialMessaging::write(SerialProtocol::Action::System::Error,
                             SerialProtocol::Error::Package::InvalidSensorType);
  }

  if (queuedSensor->getParameterByteCount() !=
      queuedPackage.getParameterCount()) {
    SerialMessaging::write(
        SerialProtocol::Action::System::Error,
        SerialProtocol::Error::Package::InvalidSensorParameters);
    return false;
  }

  if (!checkCrc8Checksum(queuedPackage))
    return true; // TODO: Error message

  queuedSensor->parseParameters(queuedPackage);

  if (arsemiArduinoCore.addSensor(queuedSensor)) {
    SerialMessaging::write(
        SerialProtocol::Action::Setup::SuccessfullyAddedSensor);
    queuedSensor->begin();
  } else
    SerialMessaging::write(SerialProtocol::Action::System::Error,
                           SerialProtocol::Error::Package::SensorCountOverflow);

  queuedPackage.reset();
  queuedSensor = nullptr;
  return true;
}

/// @brief Check if the checksum is correct, otherwise send an error message
/// over serial
/// @param crc8Checksum the checksum of the package
/// @param package package to be checked
/// @return true if calculated checksum is similar to crc8Checksum, otherwise
/// false
bool MessageParsing::checkCrc8Checksum(SerialPackage &package) {
  uint8_t nextCrc8Checksum = queuedPackage.Crc8;
  uint8_t calculatedCrc8Checksum = SerialMessaging::CRC8(queuedPackage);

  // uint8_t *serializedPackage = package.Serialize();

  if (nextCrc8Checksum != calculatedCrc8Checksum) {
    uint8_t errorPackage[4] = {SerialProtocol::Action::System::Error,
                               SerialProtocol::Error::Package::InvalidChecksum,
                               nextCrc8Checksum, calculatedCrc8Checksum};
    SerialMessaging::write(errorPackage, 4);
    // SerialMessaging::write(SerialProtocol::Action::System::Error,
    //                        SerialProtocol::Error::Package::InvalidChecksum);
    return false;
  }
  return true;
}